hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala

```diff
@@ -1044,16 +1045,30 @@ class DependentEmitFunctionBuilder[F](
       field
     })
 
-  override def addLiteral(v: Any, t: PType): Code[_] = {
+  override def addLiteral(v: Any, t: PType): PValue = {
     assert(v != null)
     literalsMap.getOrElseUpdate(t -> v, {
       val fromParent = parentcb.addLiteral(v, t)
-      val ti: TypeInfo[_] = typeToTypeInfo(t)
-      val field = newDepFieldAny(fromParent)(ti)
-      field
+      newDepPField(fromParent.get)
     })
   }
 
+  def newDepPField(pc: PCode): PValue = {
+    val ti = typeToTypeInfo(pc.pt)
+    val field = newPField(pc.pt)
+    dep_apply_method.setFields += { (obj: lir.ValueX) =>
+      val code = pc.code
+      // XXX below assumes that the first settable is the 'base' of the PSettable
+      val baseField = field.settableTuple()(0).asInstanceOf[ThisFieldRef[_]]
+      code.end.append(lir.putField(className, baseField.name, ti, obj, code.v))
+      // FIXME need to initialize other potential settables in the PSettable here
```

Comment:
This is still definitely the part I feel least confident in. I'm not sure this code is ever used, so it continues to pass tests, but I would feel much better if I could store everything properly.

Reply:
@cseed This is the only bit I where I specifically need feedback from you, in dependent methods.

Reply:
yeah, I agree this isn't ideal. I think we'll break this when we extend PCode, but it seems fine for now.